File: solutions/c/intergalactic-transmission/1/intergalactic_transmission.c
#include "intergalactic_transmission.h"

uint8_t parity(uint8_t byte){
    uint8_t parity = 0;
    for (uint8_t i = 0; i < 7; i++)
        parity ^= (byte >> (7-i)) & 1;
    byte &= 254;
    byte |= parity;
    return byte; 
}

uint8_t test(const uint8_t *message, int message_length){
    uint8_t odd = 0;
    for (int i = 0; i < message_length; i++, odd = 0){
        for (uint8_t j = 0; j < 7; j++)
            odd ^= (message[i] >> (7 - j)) & 1;
        if((message[i] & 1) != odd) return 1;
    }
    return 0;
}

int transmit_sequence(uint8_t *buffer, 
                      const uint8_t *message,
                      int message_length){
    if (!message_length) return 0;
    int j = 0;
    uint8_t byte;
    for (int i = 0; i < message_length; i++, j++){
        if (i%7 == 0)
            buffer[j] = parity(message[i]);
        else{
            byte = message[i] >> (i%7);
            byte |= message[i-1] << (8-i%7);
            buffer[j] = parity(byte);
            if (i%7 == 6)
                buffer[1+j++] = parity(message[i] << 1);
        }
    }
    if (j%7 != (message_length/7)%7){
        byte = message[message_length - 1] << (8 - j%8);
        buffer[j++] = parity(byte);
    }
return j;
}

int decode_message(uint8_t *buffer, 
                   const uint8_t *message, 
                   int message_length){
    if (!message_length) return 0;
    if (test(message, message_length))
        return WRONG_PARITY;    
    int j = 0;
    uint8_t byte;
    for (int i = 0; i < message_length; i++, j++){
        if (i && j%7 == 0) i++;
        byte = (message[i] & 254) << j%7;
        if (i + 1 != message_length)
            byte |= message[i+1] >> (7 - j%7);
        buffer[j] = byte;
    }
    return j - 1;
}

File: solutions/julia/intergalactic-transmission/1/intergalactic-transmission.jl
function parity(byte)
    folded_bits = byte
    i = 4
    while i > 0
        folded_bits ⊻= folded_bits >> i
        i >>= 1
    end
    folded_bits &= 1
    byte ⊻= folded_bits
    byte
end

function test(message)
    for byte in message
        folded_bits = byte
        j = 4
        while j > 0
            folded_bits ⊻= folded_bits >> j
            j >>= 1
        end
        if folded_bits & 1 == 1; return 1; end
    end
    0
end

function get_transmit_sequence(message)
    if isempty(message); return []; end
    buffer = UInt8[]
    for i in eachindex(message)
        if (i - 1) % 7 == 0
            push!(buffer, parity(message[i]))
        else
            byte = message[i] >> ((i - 1) % 7)
            byte |= message[i - 1] << (8 - (i - 1) % 7)
            push!(buffer, parity(byte))
            if (i - 1) % 7 == 6
                push!(buffer, parity(message[i] << 1))
            end
        end
    end
    j = length(buffer)
    if j % 7 != (length(message) ÷ 7) % 7
        byte = message[end] << (8 - j % 8)
        push!(buffer, parity(byte))
    end
    buffer
end

function decode_sequence(received_seq)
    if isempty(received_seq); return []; end
    if test(received_seq) != 0; error("Invalid parity"); end
    decoded_length = (length(received_seq) * 7) ÷ 8
    buffer = UInt8[]
    i = 1
    for j in 1:decoded_length
        if j > 1 && (j - 1) % 7 == 0
            i += 1
        end
        byte = (UInt8(received_seq[i]) & 0xfe) << ((j - 1) % 7)
        if i + 1 <= length(received_seq)
            byte |= UInt8(received_seq[i + 1] >> (7 - (j - 1) % 7))
        end
        push!(buffer, byte)
        i += 1
    end
    buffer
end